File: project/core/lib/common/src/main/java/vdi/core/health/Dependency.java
package vdi.core.health;

import java.util.Map;

public interface Dependency {
  String getName();
  String getProtocol();
  String getHost();
  int getPort();
  Map<String, Object> getExtraFields();

  enum Status {
    OK,
    NOT_OK,
    UNKNOWN
  }

  Status checkStatus();

  default String urlString() {
    return this.getProtocol().isBlank()
      ? String.format("%s:%d", this.getHost(), this.getPort())
      : String.format("%s://%s:%d", this.getProtocol(), this.getHost(), this.getPort());
  }
}

File: project/core/lib/common/src/main/kotlin/vdi/core/health/Dependency.kt
package vdi.core.health

interface Dependency {
  val name: String
  val protocol: String
  val host: String
  val port: UShort
  val extra: Map<String, Any>

  enum class Status {
    Ok,
    NotOk,
    Unknown,
  }

  fun checkStatus(): Status

  fun urlString() =
    if (protocol.isNotBlank())
      "$protocol://$host"
    else
      host
}
